### PR TITLE
Recommend using the 0x prefix in App.config

### PR DESCRIPTION
**Pull request by cshung**
## Summary

Update the GC Configuration page to recommend using 0x prefix in `App.config` to avoid a known unfortunate bug.

@dotnet/gc

### CODE REVIEW

File: docs/core/runtime-config/garbage-collector.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ Settings are arranged into groups on this page. The settings within each group a
 > - These settings can also be changed dynamically by the app as it's running, so any run-time settings you set may be overridden.
 > - Some settings, such as [latency level](../../standard/garbage-collection/latency.md), are typically set only through the API at design time. Such settings are omitted from this page.
 > - For number values, use decimal notation for settings in the *runtimeconfig.json* file and hexadecimal notation for environment variable settings. For hexadecimal values, you can specify them with or without the "0x" prefix.
```
Comment (gewarren) [inline docs/core/runtime-config/garbage-collector.md:17]:
```suggestion
> - For number values, use decimal notation for settings in the *runtimeconfig.json* file and hexadecimal notation for environment variable and *app.confg* settings. For hexadecimal values, you can specify them with or without the "0x" prefix. However, for *app.config* settings, we recommend always adding the "0x" prefix. The prefix ensures that the number is interpreted as hexadecimal and not decimal in all versions of .NET Framework.
```

File: docs/core/runtime-config/garbage-collector.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ Settings are arranged into groups on this page. The settings within each group a
...
+> - Due to an unfortunate bug, while numbers in the App.config without the 0x prefix is interpreted as decimal in .NET framework 4.7.2, numbers for GC config only are interpreted as hex in .NET framework 4.8 or above. To avoid issues, we recommend always specifying numbers as hex with the 0x prefix in App.config, that way it will work on any .NET framework versions.
```
Comment (gewarren) [inline docs/core/runtime-config/garbage-collector.md:18]:
```suggestion
```

**Review by gewarren — COMMENTED** [review]
What do you think about combining it with the previous bullet point? It makes sense to me.

**Comment by cshung** [comment]
Thanks for the rewrite, but I am afraid the rewrite is less clear though. 

If I were a customer trying to set the `GCHeapCount` the natural base to use is decimal and I would write 12 there. If I read your message, it isn't clear why I would want to change it to hex and add the 0x prefix. It is more obscure that way and doesn't make anything better. The one with decimal notation will work just fine (on 4.7.2.) and mostly fine on 4.8 (but got misinterpreted as 18, which I wouldn't notice until I really look into it in the debugger).

That's a trap that we want customers to avoid. The reason I added this note is to tell customers about this trap so they can avoid it.

I don't have a strong preference for emphasizing it is a bug, but the message needs to redirect the user to use the 0x hex notation even when it is unnatural to do so.

File: docs/core/runtime-config/garbage-collector.md (hunk 1 above)
Comment (gewarren) [reply docs/core/runtime-config/garbage-collector.md:17]:
```suggestion
> - For number values, use decimal notation for settings in the *runtimeconfig.json* file and hexadecimal notation for environment variable and *app.config* settings. For hexadecimal values, you can specify them with or without the "0x" prefix. However, for *app.config* settings, we recommend always adding the "0x" prefix. In certain versions of .NET Framework, if you omit the prefix on a number value, it's interpreted as a decimal value.
```

**Comment by gewarren** [comment]
@cshung I made a different suggestion. Is that more clear?

**Comment by Maoni0** [comment]
so, I already heard feedback from customers that the description of how to specify configs sounded confusing. with this additional note it's even more info to take just to specify a config. I think if I were a customer I'd appreciate something less to read. how about changing the description to something like a table? please let me know if this seems clearer to both of you.



| config location    | .net versions this location applies to | formats  | how it's interpreted                                         |
| ------------------ | -------------------------------------- | -------- | ------------------------------------------------------------ |
| app.config         | .NET Framework                         | 0xn   | n is interpret as a hex value  (also see note below)                              |
| env var            | .NET Framework, .NET Core              | 0xn or n | n is interpret as a hex n is interpret as a hex value number in either format |
| runtimeconfig.json | .NET Core                              | n        | n is interpreted as a decimal value.                         |

Note that you could specify a value with the 0x prefix for app.config but we do not recommend that. Due to an unfortunate bug, on .NET Framework 4.8+ it's interpreted as hex but on previous versions of .NET Framework it's interpreted as decimal. To avoid having to change your config please always use the 0x prefix when specifying a value in your app.config.

An example, if I want to specify 12 heaps for GCHeapCount, this is how I would specify it for an executable named a.exe [give the example how to specify that in all 3 ways]

**Comment by gewarren** [comment]
I think it's a great idea @Maoni0!

**Comment by Maoni0** [comment]
looks good. my text about the example sounds too casual for official doc - @gewarren could you please help us make it more suitable for official documentation?

I'd be totally fine if we left this in the GC doc page just because some of this is specific to GC (eg the 4.8 bug). and it's nice to have just one page to go to.

**Comment by gewarren** [comment]
How does this look? [Preview](https://review.docs.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector?branch=pr-en-us-28943#ways-to-specify-the-configuration)

**Review by Maoni0 — APPROVED** [review]

**Pull request merged**